### ISSUE TITLE
Package-specific make/cmake arguments

### ISSUE THREAD

**Issue by scpeters** [catkin/catkin_tools#205]
It would be great to be able to specify make and cmake arguments on a per-package basis. For example, I always build bullet with cmake arguments `-DBUILD_SHARED_LIBS=ON` and `-DUSE_DOUBLE_PRECISION=ON`. I currently just send those variables to all packages in my workspace, which has worked so far, but it's not a robust solution.

It would also be great if these arguments could be remembered in the configuration file, so I don't have to specify them all the time when I do `catkin build`.

**Comment by jbohren** [comment]
I've thought about this often, and there are definitely motivating use cases. I think after we have build warning capture and per-package cleaning, we should talk about this and the catkin_tools roadmap.

**Comment by adolfo-rt** [comment]
+1

I can also see this useful when one wants to build a package against a dependency installed in a non-standard location (`/opt`, `/home/xxx`, ...), and not against a system install which might be present in a default location (`/usr/...`).

**Comment by jbohren** [comment]
It could look like this when you view the config summary:

```
-----------------------------------------------------------
Profile:                     default
Extending:          [cached] /opt/ros/hydro
Workspace:                   /home/user/ws/foobar
Source Space:       [exists] /home/user/ws/foobar/src
Build Space:        [exists] /home/user/ws/foobar/build
Devel Space:        [exists] /home/user/ws/foobar/devel
Install Space:     [missing] /home/user/ws/foobar/install
DESTDIR:                     None
-----------------------------------------------------------
Isolate Develspaces:         False
Install Packages:            False
Isolate Installs:            False
-----------------------------------------------------------
Whitelisted Packages:        None
Blacklisted Packages:        None
-----------------------------------------------------------
Internal Job Server:         True
CMake Args:                  No Global Args
 - my_pkg: -DFoo=1 -DBar=2
 - my_otherpkg: -DFoo=2 -DBar=1
Make Args:                   None
Make Args (Catkin):          None
-----------------------------------------------------------
Workspace configuration appears valid.
-----------------------------------------------------------
```

**Comment by scpeters** [comment]
+1 to the CLI and summary proposal by @jbohren

**Comment by NikolausDemmel** [comment]
Yes, this would be really nice to have. The use case I currently have is that I would like to build one package as Debug, and everything else as Release.

+1 also for the proposed interface.

**Comment by ojura** [comment]
+1 for this also. Is this still a possibility?

**Comment by wjwwood** [comment]
We would like to have it, but I don't know when it would be implemented.

**Comment by mikepurvis** [comment]
I'm interested in this too, although my use case with bundle builds is a workspace of multiple hundreds of packages, where there are a handful of one-off flag modifications, but mostly it's two large groups of configurations. Because of this, the proposed CLI config output shown above would be pretty unwieldy— I'd need a flag to suppress showing that and then I could grab the .catkin directory as a build artifact if it needed to be inspected.

I wonder about leveraging the existing "profile" scheme in catkin_tools, so you'd basically set up your profiles with the flags you want, then assign packages to the profiles, and do a multi-profile build.

**Comment by dirk-thomas** [comment]
I just want to mention this here since it might safe/avoid some effort: the prototype of the [unified build tool[(http://design.ros2.org/articles/build_tool.html) I am working on in my spare time already supports that use case. The package specific arguments / configuration can be provided using different approaches: a config file places anywhere e.g. in the workspace, a file places within the package itself, etc.

**Comment by Timple** [comment]
> already supports that use case

Would you mind sharing the syntax for this in colcon?
I can't seem to find it [here](https://colcon.readthedocs.io/en/released/reference/verb/build.html)

**Comment by tonynajjar** [comment]
> > already supports that use case
> 
> Would you mind sharing the syntax for this in colcon? I can't seem to find it [here](https://colcon.readthedocs.io/en/released/reference/verb/build.html)

+1 on this. Is it possible with colcon?
